### PR TITLE
Issue #4369: Investigate how to map XPath expressions onto AST nodes

### PR DESCRIPTION
**Pull request by timurt**
https://github.com/checkstyle/checkstyle/issues/4369

### CODE REVIEW

**Comment by rnveach** [comment]
@timurt Please run `mvn clean verify` before pushing new changes and make sure it passes.
A lot of CI failures will be picked up with that command.

**Comment by timurt** [comment]
@rnveach yeah I know, I just wanted to share my implementation with you, so you can comment on code. 
Is it better to share my forked repository or correct errors and make pr?

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+     * @param detailAST reference to {@code DetailAST}
+     */
+    public ElementNode(int tokenType, Node parent, Node document, DetailNode detailNode, DetailAST detailAST) {
+        super(tokenType);
+        this.parent = parent;
+        this.document = document;
+        this.detailNode = detailNode;
+        this.detailAST = detailAST;
+
+        if (detailAST.branchContains(TokenTypes.IDENT)) {
```
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: looks like a hack. How are you going to support other tokens?

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import net.sf.saxon.type.Type;
+
+/**
+ * Represents element node of Xpath-tree.
+ *
+ * @author Timur Tibeyev
+ */
+public class ElementNode extends AbstractNode {
+    /** Represents value of TokenTypes#IDENT. */
+    private String image;
```
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: 'image` is not a correct name to represent a node type. PMD uses that name due to the fact that their API work with that name, and that is a historical problem of their design.

**Comment by rnveach** [comment]
@timurt IMO, if you wish to share code and receive specific comments, you should share your repo/branch with your mentor. @MEZk will know more about this implementation than anyone else.
He should be able to directly comment there on github.

I don't recommend using the PR system because we are already flooded with PRs and it will be harder for us to manage what is ready for a 'final' review and what isn't.

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * Returns display name.
+     * @return display name
+     */
+    public String getDisplayName() {
+        return detailAST.getText();
+    }
+
+    public String getAttributeValue(String s, String s1) {
```
Comment (MEZk) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: what does `s1` mean? Why it is so special? What is the difference between `s` and `s1`? 
The names of the parameters should be changed to explain their purposes.

File: src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        final FileContents contents = new FileContents(text);
+
+        DetailAST rootAST = TreeWalker.parse(contents);
+        DetailNode detailNode = null;
+
+        RootNode rootNode = new RootNode();
+        ElementNode classDefinition = new ElementNode(TokenTypes.CLASS_DEF, rootNode, rootNode, detailNode, rootAST.getNextSibling());
+
+        XPathEvaluator xpathEvaluator = new XPathEvaluator();
+        String xpath = "/CLASS_DEF[@Image='TestAst']";
```
Comment (MEZk) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java]: @romani @rnveach @timurt 
I'm not OK with the name of the attribute. What does it mean? DetailAST has nothing in common with Image. It is not the class field or getter/setter. If we want to refer to a class with the name `TestAst` the xpath-query should be:
```
//CLASS_DEF[@Name='TestAst']
```

or

```
//CLASS_DEF[@Text='TestAst']
```

as actually we refer to the field with the name text which is a part of CommonAST class.

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java]: Taken from romani's comments at https://github.com/checkstyle/checkstyle/issues/4369#issuecomment-307098907,

> It should be //CLASS_DEF[@text='TestClass'] , TokenType is used as name ,all attributes are properties from DetailAST.

Comment (MEZk) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java]: @rnveach 
Yes, it is OK. 

@timurt 
Please, fix.

**Comment by MEZk** [comment]
@timurt 
Ping

File: src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    public void testOne() throws XPathException, IOException, TokenStreamException, RecognitionException {
+        final FileText text = new FileText(new File("src/test/resources/com/puppycrawl/tools/checkstyle/api/TestAst.java"),
+                System.getProperty("file.encoding", "UTF-8"));
+        final FileContents contents = new FileContents(text);
+
+        DetailAST rootAST = TreeWalker.parse(contents);
+        DetailNode detailNode = null;
+
+        RootNode rootNode = new RootNode();
+        ElementNode classDefinition = new ElementNode(TokenTypes.CLASS_DEF, rootNode, rootNode, detailNode, rootAST.getNextSibling());
```
Comment (timurt) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java]: I don't really understand how DetailNode relate to DetalAST, is it one of the possible children?

File: src/test/resources/com/puppycrawl/tools/checkstyle/api/TestAst.java [hunk 6]
```diff
@@ -0,0 +1,4 @@
+package com.puppycrawl.tools.checkstyle.api;
```
Comment (timurt) [inline src/test/resources/com/puppycrawl/tools/checkstyle/api/TestAst.java]: I will move it to another folder

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 2 above)
Comment (timurt) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: ok, my bad

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/Node.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+import net.sf.saxon.tree.wrapper.SiblingCountingNode;
+import net.sf.saxon.tree.wrapper.VirtualNode;
+
+/**
+ * Node interface
+ *
+ * @author Timur Tibeyev
+ */
+public interface Node extends VirtualNode, SiblingCountingNode {
```
Comment (timurt) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/Node.java]: I think we don't need it

File: src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java (hunk 1 above)
Comment (timurt) [inline src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java]: it is not hack, as I understood `IDENT` is used to identify names of classes, methods, parameters and etc.
So this part of code extracts name of the current AST element and make it possible to query like `@text = 'SomeName' ` using `getAttribute(namespace, localpart)` method

**Comment by timurt** [comment]
@MEZk 
I am here, working on more deep queries, creating new tests

File: src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java (hunk 5 above)
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java]: `DetailNode` is for Javadoc ANTLR. `DetailAST` is for Java ANTLR.

Comment (timurt) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/XpathMapperTest.java]: @rnveach Does it mean they don't have references to each other and we should be able to create two different XPATH trees one for javadoc and one for java

**Pull request closed without merging**